lib/3scale/backend/alerts.rb

```diff
@@ -45,43 +50,82 @@ def key_current_id
       FIRST_ALERT_BIN = ALERT_BINS.first
       RALERT_BINS     = ALERT_BINS.reverse.freeze
 
-      def can_raise_more_alerts?(service_id, app_id)
-        allowed_bins = allowed_set_for_service(service_id).sort
+      # This class is useful to reduce the amount of information that we need to
+      # fetch from Redis to determine whether an alert should be raised.
+      # In summary, alerts are raised at the application level and we need to
+      # wait for 24h before raising a new one for the same level (ALERTS_BIN
+      # above).
+      #
+      # This class allows us to check all the usage limits once and then not
+      # check all of them again (just the ones in the report job) until:
+      # 1) A specific alert has expired (24h passed since it was triggered).
+      # 2) A new alert bin is enabled for the service.
+      class UsagesChecked
+        extend KeyHelpers
+        extend StorageHelpers
+        include Memoizer::Decorator
+
+        def self.need_to_check_all?(service_id, app_id)
+          !storage.exists(key_usage_already_checked(service_id, app_id))
+        end
+        memoize :need_to_check_all?
 
-        return false if allowed_bins.empty?
+        def self.mark_all_checked(service_id, app_id)
+          ttl = ALERT_BINS.map do |bin|
+            ttl = storage.ttl(key_already_notified(service_id, app_id, bin))
 
-        # If the bin with the highest value has already been notified, there's
-        # no need to notify anything else.
-        not notified?(service_id, app_id, allowed_bins.last)
-      end
+            # Redis returns -2 if key does not exist, and -1 if it exists without
+            # a TTL (we know this should not happen for the alert bins).
+            # In those cases we should just set the TTL to the max (ALERT_TTL).
+            ttl >= 0 ? ttl : ALERT_TTL
+          end.min
 
-      def utilization(app_usage_reports)
-        max_utilization = -1.0
-        max_record = nil
-        max = proc do |item|
-          if item.max_value > 0
-            utilization = item.current_value / item.max_value.to_f
-
-            if utilization > max_utilization
-              max_record = item
-              max_utilization = utilization
-            end
-          end
+          storage.setex(key_usage_already_checked(service_id, app_id), ttl, '1'.freeze)
+          Memoizer.clear(Memoizer.build_key(self, :need_to_check_all?, service_id, app_id))
```

Comment:
Rather than clearing you could just set the new memoized value for `:need_to_check_all?` (would avoid one command per memorization cycle, so no big deal).

Reply:
That would increase complexity a bit in my opinion.
I don't think there's a clear winner in this case, bud I'd rather leave it as it is now.